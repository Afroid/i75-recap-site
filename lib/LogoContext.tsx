"use client"; // This module is purely client-side: allows useState/useEffect
import { createContext, useContext, useState, useEffect } from "react";

// List of possible logos. One is displayed at random, but
// it always starts with "/default-logo.png" to avoid hydration mismatch.
const logos = [
  "/default-logo.png",
  "/default-logo-alt.png",
  "/default-RWB.png",
];

// Create a context whose default value matches our initial state.
// This ensures SSR output will match the client-side initial render.
const LogoContext = createContext("/default-logo.png");

export function LogoProvider({ children }: { children: React.ReactNode }) {
  // This starts with a fixed default so SSR + initial hydration match
  const [logo, setLogo] = useState<string>("/default-logo.png");

  // Once it's mounted, this will pick a random GIF exactly once
  useEffect(() => {
    const choice = logos[Math.floor(Math.random() * logos.length)];
    setLogo(choice);
  }, []); // empty deps: run only on first mount

  return (
    <LogoContext.Provider value={logo}>
      {children}
    </LogoContext.Provider>
  );
}

export function useLogo() {
  return useContext(LogoContext);
}
