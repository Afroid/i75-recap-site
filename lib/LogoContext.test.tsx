import React from "react";
import { render, renderHook, waitFor } from "@testing-library/react";
import { LogoProvider, useLogo } from "./LogoContext";

// A small component that reads from your LogoContext
function TestLogo() {
  const logo = useLogo();
  return <img data-testid="logo" src={logo} alt="Logo" />;
}

describe("LogoContext / useLogo", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("defaults to the initial logo during SSR/hydration", () => {
    // before renderHook, make Math.random return 0 so
    // logos[0] === "/default-logo.png"
    jest.spyOn(Math, "random").mockReturnValue(0);

    const { result } = renderHook(() => useLogo(), {
      wrapper: ({ children }) => <LogoProvider>{children}</LogoProvider>,
    });

    // Assert the initial state is always "/default-logo.png"
    expect(result.current).toBe("/default-logo.png");

    jest.restoreAllMocks();
  });

  it("picks one of the known logos after mount", async () => {
    const { result } = renderHook(() => useLogo(), {
      wrapper: ({ children }) => <LogoProvider>{children}</LogoProvider>,
    });

    // Assert state can be any of the three random options
    await waitFor(() => {
      expect([
        "/default-logo.png",
        "/default-logo-alt.png",
        "/default-RWB.png",
      ]).toContain(result.current);
    });
  });

  it("matches a snapshot of the markup with default-logo.png", () => {
    // Before renderHook, make Math.random return 0 anad 1 on first and second calls, respectively.
    // logos[0] === "/default-logo.png"
    // logos[2] === "/default-RWB.png"
    const mathRandomMock = jest.spyOn(Math, "random")
      .mockReturnValueOnce(0)      // for the first mount
      .mockReturnValueOnce(0.8);   // for the second mount

    const { asFragment, rerender } = render(
      <LogoProvider key="first">
        <TestLogo />
      </LogoProvider>
    );

    // Before effect, one state
    expect(asFragment()).toMatchSnapshot(); // Snapshot 1 uses "/default-logo.png"

    // Re-render with new src
    // Adding keys forces a real remount on the second render
    // Old tree is torn down to build a new one, which re-runs useEffect()
    rerender(
      <LogoProvider key="second">
        <TestLogo />
      </LogoProvider>
    );

    // logos[1] === "/default-RWB.png"
    expect(asFragment()).toMatchSnapshot(); // Snapshot 2 uses "/default-RWB.png"

    // Restore mathRandomMock
    mathRandomMock.mockRestore();
  });
});
