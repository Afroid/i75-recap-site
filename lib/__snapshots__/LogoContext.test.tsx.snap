// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoContext / useLogo matches a snapshot of the markup with default-logo.png 1`] = `
<DocumentFragment>
  <img
    alt="Logo"
    data-testid="logo"
    src="/default-logo.png"
  />
</DocumentFragment>
`;

exports[`LogoContext / useLogo matches a snapshot of the markup with default-logo.png 2`] = `
<DocumentFragment>
  <img
    alt="Logo"
    data-testid="logo"
    src="/default-RWB.png"
  />
</DocumentFragment>
`;
